Native Instruments Traktor Kontrol S2 MK3 mapping

mi01 opened the pull request:
Hello,
i created a USB-HID controller mapping for the Native Instruments Traktor Kontrol S2 MK3. I think its complete now, so i would like to contribute it officially here. There is a [wiki page](https://www.mixxx.org/wiki/doku.php/native_instruments_traktor_kontrol_s2_mk3) and a [thread](https://www.mixxx.org/forums/viewtopic.php?f=7&t=12999) in the forums also with additional infos. I'm looking forward to get some feedback.

mi01 committed:
Traktor Kontrol S2MK3: Initial mapping script

mi01 committed:
Traktor Kontrol S2MK3: VuMeter handling

mi01 committed:
Traktor Kontrol S2MK3: Microphone button

mi01 committed:
Traktor Kontrol S2MK3: Samplers

mi01 committed:
Traktor Kontrol S2MK3: Remove HID specs from source file

mi01 committed:
Traktor Kontrol S2MK3: Added shift-functions for play, cue and sync

mi01 committed:
Traktor Kontrol S2MK3: Avoid spamming bus with VuMeter

mi01 committed:
Traktor Kontrol S2MK3: Better sampler handling

mi01 committed:
Traktor Kontrol S2MK3: Improved track selection & bugfixes

mi01 committed:
Traktor Kontrol S2MK3: Reverse and slip mode

mi01 committed:
Traktor Kontrol S2MK3: Better beatjump and loop handling

mi01 committed:
Traktor Kontrol S2MK3: GRID and preperation button

mi01 committed:
Traktor Kontrol S2MK3: JS code refactoring

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,980 @@
+    };
+};
+
+var TraktorS2MK3 = new TraktorS2MK3();
```
I'm not sure if it's a good idea to overwrite a class with an instance of itself. This might work, but could break in the future. Maybe just chose another name?

mi01 replied:
You are right, i will change this back to this: `var TraktorS2MK3 = new function () { ... }`. I separated class definition and instantiation because of the code checking tools giving warnings here. Since there is only one instance of the class this might be not as confusing as using a different name.

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,980 @@
+    var vuKeys = Object.keys(TraktorS2MK3.vuMeterThresholds);
+    for (var i = 0; i < vuKeys.length; ++i) {
+        // Avoid spamming HID by only sending last LED update
+        var last = (i === vuKeys.length - 1);
```
This doesn't look right. Are you substracting 1 from a boolean? Just do:

```javascript
var value = ((i === vuKeys.length) ? 0 : -1);
```

mi01 replied:
Since the subtraction operator has a higher precedence than the equality operator this line is ok i think. The variable last will be true in the last iteration of the loop. But i will add some additional parentheses to make this more obvious.

Holzhaus replied:
Yes, that'd be nice.

Holzhaus reviewed (changes requested):
Some minor issues, otherwise looks good to me.

mi01 committed:
Traktor Kontrol S2MK3: Fix review problems and remove raw data sending

Holzhaus commented:
@Be-ing ping

Be-ing commented:
Thank you for helping out with the code review @Holzhaus. Reviewing the code is helpful, however my bigger concern with controller mappings is whether the UX design makes sense and works with how Mixxx is designed. This is why it is essential for a wiki page to be written before review. The contents of the wiki page can be compared with the code to verify that the code does what it is supposed to do (as much as can be done without the reviewer having the hardware). Also, the wiki page can be edited for language clarity/organization/style.

This takes some work and so far I have been the only one doing this. I am overwhelmed keeping up with mappings and trying to make releases happen. If you could help review mappings for UX design, that would be really helpful!

Holzhaus commented on res/controllers/Traktor Kontrol S2 MK3.hid.xml:
```diff
@@ -0,0 +1,19 @@
+    <name>Traktor Kontrol S2 MK3</name>
+    <author>Michael Schmidt</author>
+    <description>HID Mapping for Traktor Kontrol S2 MK3</description>
+    <wiki>https://www.mixxx.org/wiki/doku.php/native_instruments_traktor_s2_mk3</wiki>
```
Please add the forum post URL, too.

mi01 replied:
Done in https://github.com/mi01/mixxx/commit/d5822606d88dbaad3ff81d4a64d15c12f2ebfb95

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,974 @@
+    } else if (value === 1 || value === true) {
+        // On value
+        ledValue = 0x7E;
+    }
```
Wouldn't this suffice?
```javascript
// Custom value for multi-colored LEDs
var ledValue = value ? 0x7E : 0x7C;
```

Holzhaus replied:
Ah ok, `value` can also have values other than `0`, `1`, `true` or `false` that are passed through without change, right?

mi01 replied:
Right, some buttons on the controller provide multiple colors for the LED. In this case the no default value is used as mentioned in the comment.

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,974 @@
+    // Reset sync button timer if active
+    if (TraktorS2MK3.syncPressedTimer[group] !== 0) {
+        engine.stopTimer(TraktorS2MK3.syncPressedTimer[group]);
+        TraktorS2MK3.syncPressedTimer[group] = 0;
```
You're resetting the timer, but you're not changing any control object. According to your wiki entry, the mapping is supposed to toggle `sync_master`.

mi01 replied:
As written in https://www.mixxx.org/wiki/doku.php/mixxxcontrols#channeln under `sync_enabled` Mixxx distinguishes between pressing the SYNC button for a long period (< 1 sec) and a short press. I just imitate this behavior in my script, since Mixxx is not checking the duration when setting the control from a controller script. This is also mentioned in the Wiki page, but i made it more clear there. The timer is started when the button is pressed. When it is released the `syncHandler()` function is checking, if the timer is still running. In this case the button was pressed shorter than one second the SYNC lock is not activated by reseting the control. But you pointed to the right spot, there was some unnecessary code in `syncHandler()`, because stopping the timer when it is already expired is useless. I fixed this in https://github.com/mi01/mixxx/commit/0ac84084748407cab4e81aefad0e6544e351c3f8

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,974 @@
+
+TraktorS2MK3.syncHandler = function (field) {
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "beatsync_phase", field.value);
```
Ok, but the shift behaviour is not really documented. In the wiki entry, it sounds like this depends on quantize, not the shift state.

mi01 replied:
I forgot this in the Wiki page, updated it now.

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,974 @@
+            }
+
+            var sync = engine.getValue(field.group, "sync_enabled");
+            engine.setValue(field.group, "sync_enabled", !sync);
```
I've seen this idiom a few times in your script, but I think you can just do:

```javascript
script.toggleControl(field.group, "sync_enabled")
```

mi01 replied:
The definition of toggleControl is quite close to my idiom:
```javascript
script.toggleControl = function(group, control) {
    engine.setValue(group, control, !(engine.getValue(group, control)))
}
```
Using this function might save a few lines of code, but I dont use any other functions from _common-controller-scripts.js_ and i need to include it in this case (in the XML file). So i would prefer to skip this change request.

Be-ing replied:
common-controller-scripts.js is included into every script automatically by Mixxx. It is like a standard library for the scripting environment. Please make use of this function instead of reinventing the wheel.

mi01 replied:
Ok, i changed this in https://github.com/mi01/mixxx/commit/1bdd1e31add3fe11768818c1cb18c4b78b419722

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,974 @@
+    engine.setValue("[Channel1]", "quantize", res);
+    engine.setValue("[Channel2]", "quantize", res);
+    TraktorS2MK3.outputHandler(res, field.group, "quantize");
+};
```
You could make this shift dependent: w/o shift: left deck, w/ shift: right deck. I'm not sure if that would improve things though, so feel free to ignore this ;-)

mi01 replied:
I think this would confuse people, since there is only one quantize button on the controller and its not obvious which deck is meant.

Holzhaus replied:
OK.

Holzhaus commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,974 @@
+TraktorS2MK3.jogTouchHandler = function (field) {
+    var deckNumber = TraktorS2MK3.controller.resolveDeck(field.group);
+    if (field.value > 0) {
+        engine.scratchEnable(deckNumber, 1024, 33.3333, 0.125, 0.125 / 8, true);
```
```suggestion
        engine.scratchEnable(deckNumber, 1024, 33 + 1/3, 0.125, 0.125 / 8, true);
```
(That's what other mappings are doing)

mi01 committed:
Traktor Kontrol S2MK3: Forums link in XML

mi01 committed:
Traktor Kontrol S2MK3: Do not stop already expired SYNC timer

mi01 committed:
Traktor Kontrol S2MK3: More precise intervals per rev

Co-Authored-By: Jan Holthuis <holthuis.jan@googlemail.com>

mi01 committed:
Traktor Kontrol S2MK3: Use toggleControl()

Be-ing commented:
It looks like the intro to the wiki page was copied and pasted:

> The Kontrol S2 MK3 is a two-channel controller with an integrated sound card. It has two integrated stereo outputs (line and 1/8“ / 3.5 mm TRS), headphone outputs (1/8” / 3.5 mm TRS) and microphone inputs (1/4“ / 6.3 mm TRS). The MK2 uses the standard HID protocol to send and receive signals from a computer, so it can work with Mixxx. The Kontrol S2 MK3 can run from USB bus power. Using the separate power supply increases the brightness of the LEDs, which is helpful for using it in daylight, and increases the volume of the headphone output. 

Can you verify that this is all correct? Does this controller have a power supply separate from the USB power?

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,965 @@
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "cue_set", field.value);
+    } else {
+        script.toggleControl(field.group, "play");
```
Oops, sorry for suggesting `script.toggleControl` for these buttons. This isn't quite correct. Set the ControlObject to 1 on button press and back to 0 on button release. `script.toggleControl` is useful for mapping button-like controls to controller components that are not buttons, for example `loop_halve`/`loop_double` to an encoder.

Be-ing commented:
I read through the wiki page and made some edits for clarity. Everything seems to be mapped in a sensible way with two small exceptions.

You could map the sampler gain knob to control the gain knobs of all the samplers together. This is an ugly hack but it's the best we can do for now. In the future I want to implement a [sampler output bus](https://bugs.launchpad.net/mixxx/+bug/1762294) that would have its own gain knob.

The way you have mapped the FX buttons in the middle of the mixer isn't all that useful without any other effects controls available on the controller. However, I don't know what else to suggest to do with them for now. Do you have any other ideas? With #1705 one of my goals is to implement something like Traktor's Mixer FX but even more powerful so you could save custom effect *chains* and assign them to the QuickEffect knobs.

mi01 committed:
Traktor Kontrol S2MK3: Sampler pregain knob

mi01 commented:
> It looks like the intro to the wiki page was copied and pasted:
> 
> > The Kontrol S2 MK3 is a two-channel controller with an integrated sound card. It has two integrated stereo outputs (line and 1/8“ / 3.5 mm TRS), headphone outputs (1/8” / 3.5 mm TRS) and microphone inputs (1/4“ / 6.3 mm TRS). The MK2 uses the standard HID protocol to send and receive signals from a computer, so it can work with Mixxx. The Kontrol S2 MK3 can run from USB bus power. Using the separate power supply increases the brightness of the LEDs, which is helpful for using it in daylight, and increases the volume of the headphone output.
> 
> Can you verify that this is all correct? Does this controller have a power supply separate from the USB power?

Indeed i copy and pasted this part of the wiki page from the predecessors page, but its correct, there is an optional power supply (15V/1.2A). But i fixed the typo now, mentioning the MK2 instead of MK3 there. You can have a look at the manual, if you like: https://www.native-instruments.com/fileadmin/ni_media/downloads/manuals/traktor/TRAKTOR_KONTROL_S2_MK3_Manual_English_1218.pdf

> I read through the wiki page and made some edits for clarity. Everything seems to be mapped in a sensible way with two small exceptions.
> 
> You could map the sampler gain knob to control the gain knobs of all the samplers together. This is an ugly hack but it's the best we can do for now. In the future I want to implement a [sampler output bus](https://bugs.launchpad.net/mixxx/+bug/1762294) that would have its own gain knob.

Yes, this seems to be the only way to give this knob a meaningful function. Added this in https://github.com/mixxxdj/mixxx/pull/2348/commits/a7e475b12d627440113f52d7f2008cc4b90e4f49 and updated the wiki page.

> The way you have mapped the FX buttons in the middle of the mixer isn't all that useful without any other effects controls available on the controller. However, I don't know what else to suggest to do with them for now. Do you have any other ideas? With #1705 one of my goals is to implement something like Traktor's Mixer FX but even more powerful so you could save custom effect _chains_ and assign them to the QuickEffect knobs. But don't count on that feature getting implemented soon.

I dont have any other idea regarding this buttons, but assigning effect units to the decks seems not completely useless to me. However it would be great to have a improved effect system in the future and i promise to update the mapping in this case.

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    TraktorS2MK3.outputHandler(field.value, field.group, "reverse");
+};
+
+TraktorS2MK3.flxHandler = function (field) {
```
Either spell out `flux` or `slip`. `flx` is easy to confuse with `fx`

mi01 replied:
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    TraktorS2MK3.outputHandler(!slip, field.group, "flx");
+};
+
+TraktorS2MK3.gridHandler = function (field) {
```
Rename to `beatgridHandler`. At first I thought this referred to the pad grid

mi01 replied:
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+
+    // If we are in samples modes light corresponding LED
+    if (TraktorS2MK3.padModeState[deck] === 1) {
+        if (key === "play_indicator") {
```
Use `play`, not `play_indicator` here

mi01 replied:
I use `play` now in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

But i discovered an issue here. In this callback function i process events of type `play` and `track_loaded`. When a track is ejected from a sampler slot, i receive an event of type `track_loaded`. After this i receive immediately two times an event of type `play` with an value of 0 for the same sampler slot, even though no track is loaded here anymore. This is very strange and i could not figure out so far, if this is a problem in my script or of Mixxx.

Be-ing replied:
Hmm, that seems like a bug in Mixxx. Does this script still work as intended? From reading the code it doesn't seem to me that the redundant event causes a problem for this script.

mi01 replied:
The script still works because i check `engine.getValue(group, "track_loaded")` before processing the `play` event now.

Be-ing replied:
Okay. I wonder if this double `play` event occurs with normal decks too or just samplers?

mi01 replied:
I tested it right now and it seems to happen also on `[ChannelX]`.

Be-ing replied:
Thanks for testing. I reported a [bug](https://bugs.launchpad.net/mixxx/+bug/1856358)

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+                // Green light on play
+                TraktorS2MK3.outputHandler(0x9E, deck, "hotcue_" + num + "_enabled");
+            } else {
+                TraktorS2MK3.outputHandler(1, deck, "hotcue_" + num + "_enabled");
```
What color does this set the LED? Please add a comment

mi01 replied:
Actually there is a comment right there already ;)

Be-ing replied:
I see the comment above for green with 0x9E, but I don't know what color 1 sets it to.

mi01 replied:
Ah ok, fixed this in https://github.com/mi01/mixxx/commit/5ecbf327afb8cfcbe569ab97d6466402c1963105

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+        return undefined;
+    }
+
+    if (!group.match(/\[Sampler[0-9]+\]/)) {
```
Use `script.samplerRegEx` for the regular expression

mi01 replied:
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+        return undefined;
+    }
+
+    var str = group.replace(/\[Sampler/, "");
```
Use the match value from the regular expression instead of doing another string operation

mi01 replied:
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+};
+
+TraktorS2MK3.lightDeck = function (switchOff) {
+
```
extra blank line

mi01 replied:
Remove this in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    this.vuLeftConnection.disconnect();
+    this.vuRightConnection.disconnect();
+    this.clipLeftConnection.disconnect();
+    this.clipRightConnection.disconnect();
```
You don't need to do this in the script. Mixxx deletes all callback connections on shutdown.

mi01 replied:
Removed them in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    }
+
+    // Dirty hack to set initial values in the packet parser
+    var data = TraktorS2MK3.toBytes("01 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 00 00 00 00");
```
The `toBytes` function seems like an overcomplication to me. Why not just write out the array literal here?

I'm also not clear why you need to pass this initial packet to `incomingData`

mi01 replied:
I removed this debug function now and use an array instead in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Passing this initial values to the HID packet parser solve an issue i have. The parser seems to react only to "changes" of its cached state. Without this line the first button press after startup is not "recognized" and no callback function is triggered, only the second button press is handled right.

mi01 committed:
Traktor Kontrol S2MK3: Implement review feedback

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    this.controller = new HIDController();
+    this.shiftPressed = { "[Channel1]": false, "[Channel2]": false };
+    this.browseState = { "[Channel1]": 0, "[Channel2]": 0 };
+    this.loopSizeState = { "[Channel1]": 0, "[Channel2]": 0 };
```
I renamed this variables to browseKnobEncoderState, loopKnobEncoderState and moveKnobEncoderState in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    this.vuRightConnection = {};
+    this.clipLeftConnection = {};
+    this.clipRightConnection = {};
+    this.vuMeterThresholds = { "vu-18": (1 / 6), "vu-12": (1 / 6 * 2), "vu-6": (1 / 6 * 3), "vu0": (1 / 6 * 4), "vu6": (1 / 6 * 5) };
```
Fixed this in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    }
+
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "cue_set", field.value);
```
Using `start_stop` in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468 now

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+};
+
+TraktorS2MK3.cueHandler = function (field) {
+    if (field.value === 0) {
```
Fixed this in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+        if (field.value) {
+            if (!TraktorS2MK3.syncPressedTimer[field.group]) {
+                // Start timer to measure how long button is pressed
+                TraktorS2MK3.syncPressedTimer[field.group] = engine.beginTimer(1000, "TraktorS2MK3.syncTimer(\"" + field.group + "\")");
```
I fixed this in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468
Furthermore i did some refactoring, for example i use one-shot timers and closures now.

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+                TraktorS2MK3.syncPressedTimer[field.group] = engine.beginTimer(1000, "TraktorS2MK3.syncTimer(\"" + field.group + "\")");
+            }
+
+            script.toggleControl(field.group, "sync_enabled");
```
I use `beatsync` here now in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+        for (var key in TraktorS2MK3.samplerHotcuesRelation[field.group]) {
+            if (TraktorS2MK3.samplerHotcuesRelation[field.group].hasOwnProperty(key)) {
+                var loaded = engine.getValue("[Sampler" + TraktorS2MK3.samplerHotcuesRelation[field.group][key] + "]", "track_loaded");
+                TraktorS2MK3.outputHandler(loaded, field.group, "hotcue_" + key + "_enabled");
```
Renamed the output keys to `pad_X` in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+};
+
+TraktorS2MK3.numberButtonHandler = function (field) {
+    if (field.value === 0) {
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+        // Samples mode
+        var sampler = TraktorS2MK3.samplerHotcuesRelation[field.group][hotcueNumber];
+        if (TraktorS2MK3.shiftPressed[field.group]) {
+            var playing = engine.getValue("[Sampler" + sampler + "]", "play_indicator");
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+            } else {
+                engine.setValue("[Sampler" + sampler + "]", "eject", field.value);
+                // Reset eject button
+                engine.setValue("[Sampler" + sampler + "]", "eject", !field.value);
```
True, updated this in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+};
+
+TraktorS2MK3.loadTrackHandler = function (field) {
+    if (field.value === 0) {
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+TraktorS2MK3.addTrackHandler = function (field) {
+    TraktorS2MK3.outputHandler(field.value, field.group, "addTrack");
+
+    if (field.value === 0) {
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+
+TraktorS2MK3.selectLoopHandler = function (field) {
+    if ((field.value + 1) % 16 === TraktorS2MK3.loopSizeState[field.group]) {
+        engine.setValue(field.group, "loop_halve", 1);
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    } else {
+        if (delta < 0) {
+            engine.setValue(field.group, "beatjump_backward", 1);
+            engine.setValue(field.group, "beatjump_backward", 0);
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

mi01 commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+    if (field.value) {
+        if (TraktorS2MK3.microphonePressedTimer === 0) {
+            // Start timer to measure how long button is pressed
+            TraktorS2MK3.microphonePressedTimer = engine.beginTimer(1000, "TraktorS2MK3.microphoneTimer()");
```
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,978 @@
+};
+
+TraktorS2MK3.activateLoopHandler = function (field) {
+    if (field.value === 0) {
```
ping

mi01 replied:
Fixed this in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

But when looking on this function again i noticed, that the current mapping is maybe not so good. There is only one button on the controller regarding loops and mapping only `reloop_toggle` and `reloop_andstop` is not usefull. I thought to distinguish here, if a loop is currently enabled or not. If yes, than this button maps `reloop_toggle` and `reloop_andstop`, otherwhise `beatloop_activate` and `beatlooproll_activate`. What do you think?

Be-ing replied:
Yes, this could be mapped better. It looks like you haven't mapped anything for pressing the other encoder, so consider that as well. It is useful to have access to `reloop_toggle` when there is no loop active though. How about:

press loop size encoder: beatloop_activate
shift + press loop size encoder: reloop_toggle

press beatjump encoder: beatlooproll_activate
shift + press beatjump encoder: reloop_andstop

mi01 replied:
I like your proposal and updated the script like this in https://github.com/mi01/mixxx/commit/86effbbcbd4faed8f214f4bd66d5430d4422a349. I will update the Wiki page now.

mi01 committed:
Traktor Kontrol S2MK3: Comment on LED color

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -0,0 +1,927 @@
+        // Light LED while pressed
+        TraktorS2MK3.outputHandler(field.value, field.group, "sync_enabled");
+    } else {
+        script.toggleControl(field.group, "beatsync");
```
Move this below into the `if (field.value)` block and use `script.triggerControl`

Be-ing replied:
Err wait, that still isn't quite right. Take a look how this is [implemented](https://github.com/mixxxdj/mixxx/blob/master/res/controllers/midi-components-0.0.js#L255) in the Components library for MIDI devices

mi01 replied:
I think it was already working correct before, but i updated the implementation like in the Components library. https://github.com/mi01/mixxx/commit/9aaf3afcfe4435b1ebd09663d4c6b44f9bdcb090

mi01 committed:
Traktor Kontrol S2MK3: Better SYNC button handling

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -275,32 +275,29 @@ TraktorS2MK3.syncHandler = function (field) {
+                TraktorS2MK3.outputHandler(1, field.group, "sync_enabled");
+            } else {
                 // Deactivate sync lock
                 engine.setValue(field.group, "sync_enabled", 0);
```
Turn off the LED here

mi01 replied:
The LED is already turned off by the callback handler for the `sync_enabled` key. But i will add a comment so this is clear.

Be-ing replied:
okay thanks for clarifying

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -275,32 +275,29 @@ TraktorS2MK3.syncHandler = function (field) {
+                }, true);
+
+                // Light corresponding LED when button is pressed
+                TraktorS2MK3.outputHandler(1, field.group, "sync_enabled");
```
This should go inside the timer callback when `sync_enabled` is set to 1.

mi01 replied:
In general i light all LEDs on the controller in case a button is pressed, so the user has some visual feedback if the button was pressed. Also here, the LED is turned off, if the button is released again, except the timer is not running anymore and sync lock is activated.

Be-ing replied:
okay

mi01 committed:
Traktor Kontrol S2MK3: Better loop handling

Be-ing commented on res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js:
```diff
@@ -461,6 +465,18 @@ TraktorS2MK3.beatjumpHandler = function (field) {
 };
 
+TraktorS2MK3.activateBeatjumpHandler = function (field) {
+    if (field.value === 0) {
```
Remove this, otherwise the rolling loop will not be exited.

mi01 replied:
Fixed this in https://github.com/mi01/mixxx/commit/13e7ef440c9dd4d9ce332ace2afe60f4929d4351

Be-ing commented:
Please rebase this on the 2.2 branch so we can include it in the next minor release of Mixxx (2.2.4).

mi01 committed:
Traktor Kontrol S2MK3: Set LEDs on startup according to current control state

mi01 committed:
Traktor Kontrol S2MK3: Better slip mode handling

mi01 committed:
Traktor Kontrol S2MK3: Fix beatloop roll

Be-ing commented:
Great work. I am glad we have a mapping for this controller soon after its release. Just rebase this on the 2.2 branch and then it will be ready to merge.

Be-ing commented:
There are still many unrelated commits from the master branch in this. Do an interactive rebase with `git rebase -i` to remove all the commits except the ones for your mapping.

mi01 commented:
Sorry, accidentally just copied your command from above and rebased to the head of the master branch before. Hope this works now. Thank you for your great feedback and support!

Be-ing commented:
Thanks for your contribution and patience with the review.

Be-ing commented:
I moved this controller into the Community Supported Mappings list on the [wiki](https://mixxx.org/wiki/doku.php/hardware_compatibility)
